feat(ledger): Updating ledger details code to display more data

TusharPardhe opened the pull request.

TusharPardhe commented:
XRPL Bounty: https://github.com/XRPLBounties/Proposals/blob/main/bounties/2%20Ledger%20Metrics.md
Discussion and Screenshots: https://github.com/XRPLBounties/Proposals/discussions/18

TusharPardhe commented:
[mvadari](https://github.com/mvadari), [ckniffen](https://github.com/ckniffen), [mayurbhandary](https://github.com/mayurbhandary), [shawnxie999](https://github.com/shawnxie999) please review

TusharPardhe commented:
@intelliot, [mvadari](https://github.com/mvadari), [ckniffen](https://github.com/ckniffen), [mayurbhandary](https://github.com/mayurbhandary), [shawnxie999](https://github.com/shawnxie999) please review

TusharPardhe commented on public/locales/en-US/translations.json:
```diff
@@ -148,6 +150,22 @@
+  "ledger_entry_type_Offer":"Offer",
+  "ledger_entry_type_PayChannel":"Pay Channel",
+  "ledger_entry_type_RippleState":"Ripple State",
+  "ledger_entry_type_SignerList":"Signer List",
```
Can add more meaningful UI names for better understanding. 
Reference: https://xrpl.org/ledger-object-types.html

TusharPardhe commented:
@ckniffen Please review

mvadari commented:
This code does not look like it will work. Did you run it and test it locally?

1) Please fill out the PR template and include screenshots of your changes.
2) Please fix the code so that tests and linting pass.
3) Please include tests for your changes.

TusharPardhe commented on src/containers/Ledger/index.js:
```diff
@@ -90,36 +90,78 @@ class Ledger extends Component {
+        transactionTypes[tx.type] += 1
+      }
+
+      if (ledgerObjTypes[tx.ledgerEntryType] === undefined) {
```
I'll look into it

golobitch replied:
This can be fairly simply done probably like this:

```
const ledgerObjectsByType = {};
for (const obj of ledgerState) {
  if (!ledgerObjectsByType[obj["LedgerEntryType"]]) {
    ledgerObjectsByType[obj["LedgerEntryType"]] = 0;
  }
  ledgerObjectsByType[obj["LedgerEntryType"]] += 1;
}
```

And also for the UI, key is the title (AccountRoot, DirectoryNode, RippleState), and value is number of this objects for this ledger.

mvadari replied:
I recommend doing ledger objects in a separate PR, since that will require a separate microservice. You'll only get the first page of ledger objects this way, instead of all of them.

TusharPardhe replied:
> I recommend doing ledger objects in a separate PR, since that will require a separate microservice. You'll only get the first page of ledger objects this way, instead of all of them.

Makes sense, for now I will add transaction types in this PR

TusharPardhe commented on src/containers/Ledger/ledger.scss:
```diff
@@ -20,49 +20,85 @@
   }
 
   .ledger-nav {
```
Sure! Let me know 😃

TusharPardhe commented on src/containers/Ledger/ledger.scss:
```diff
@@ -144,18 +180,20 @@
           margin-bottom: 20px;
-          text-align: right;
+
+          @include for-size(desktop-up){
```
> Because of the sheer number of metrics this needs to be revisited.
> 
> 
> 
> We thinking flex-box with wrap on and a gap of 48px.

I'll have to play around with css before agreeing to your point

TusharPardhe commented:
> This code does not look like it will work. Did you run it and test it locally?
> 
> 
> 
> 1) Please fill out the PR template and include screenshots of your changes.
> 
> 2) Please fix the code so that tests and linting pass.
> 
> 3) Please include tests for your changes.

I'll add in screenshots and update the tests asap.

ckniffen commented on src/containers/Ledger/index.js:
```diff
@@ -90,36 +90,78 @@ class Ledger extends Component {
+        failedTxCount += 1
+      }
+
+      if (transactionTypes[tx.type] === undefined) {
```
DEX, Account, Payment, NFT, XChain, and Pseduo-TX

TusharPardhe replied:
Great! I will update it by tomorrow, will ping you once done

TusharPardhe replied:
DEX - Offer,
ACCOUNT - AccountRoot, RippleState, SignerList, 
PAYMENT - Check, DepositPreauth, Escrow,  PayChannel, 
NFT - NFTokenOffer, NFTokenPage, 
XCHAIN - Amendments, DirectoryNode, FeeSettings, LedgerHashes, NegativeUNL, Ticket
PSEUDO-TX -

I have categorised ledger objects based on your comments. I want to know which ledger obj will be considered as a pseudo 

Refer to this link: https://xrpl.org/ledger-object-types.html

ckniffen replied:
Sorry those were the groupings for transactions.  The objects though we can show each type.  That needs to be the total number of objects.

TusharPardhe replied:
Ahh okay now I get it.
1. Total Transactions
2. Transaction Types:  DEX, Account, Payment, NFT, XChain, and Pseduo-TX
3. Total Ledger Objects
4. Ledger Objects Types: https://xrpl.org/ledger-object-types.html

I will have to think of design for this as it'll all get clustered

TusharPardhe commented:
> > This code does not look like it will work. Did you run it and test it locally?
> > 
> > 1. Please fill out the PR template and include screenshots of your changes.
> > 2. Please fix the code so that tests and linting pass.
> > 3. Please include tests for your changes.
> 
> I'll add in screenshots and update the tests asap.

@mvadari Attaching screenshots

<img width="917" alt="Screenshot 2022-09-20 at 10 30 34 pm" src="https://user-images.githubusercontent.com/31487192/191368501-4821d4c7-8058
<img width="1496" alt="Screenshot 2022-09-20 at 10 29 53 pm" src="https://user-images.githubusercontent.com/31487192/191368516-56bbeb95-dde1-4eb4-bdc3-d5e4c8f1064c.png">
-4e86-93fc-805a9bba9f09.png">

TusharPardhe commented:
<img width="1427" alt="Screenshot 2022-09-21 at 3 22 22 pm" src="https://user-images.githubusercontent.com/31487192/191530811-2f8dcdef-0eda-4883-914e-603f9d7d6087.png">
<img width="600" alt="Screenshot 2022-09-21 at 3 22 43 pm" src="https://user-images.githubusercontent.com/31487192/191530825-c7615acd-6c98-4497-96ee-d5962902185c.png">


@mvadari @ckniffen I have updated the code and added tx types as suggested.

mvadari commented on src/containers/Ledger/test/Ledger.test.js:
```diff
@@ -60,11 +60,11 @@ describe('Ledger container', () => {
     expect(summary.length).toBe(1)
     expect(summary.find('.ledger-cols').length).toBe(1)
-    expect(summary.find('.ledger-col').length).toBe(3)
+    expect(summary.find('.ledger-col').length).toBeGreaterThanOrEqual(4)
```
This should check for the exact number of columns

mvadari commented on public/locales/en-US/translations.json:
```diff
@@ -160,6 +162,11 @@
   "transaction_EnableAmendment": "Enable Amendment",
   "transaction_SetFee": "Set Fee",
   "transaction_UNLModify": "UNL Modify",
+  "dex_transaction": "Dex transaction",
```
These should all be plural and consistent with capitalization

TusharPardhe replied:
done

TusharPardhe commented on src/containers/Ledger/index.js:
```diff
@@ -90,36 +90,109 @@ class Ledger extends Component {
     const nextIndex = LedgerIndex + 1
     const date = new Date(data.close_time)
+    const transactions = data.transactions || []
+    const transactionTypes = {
```
Yup, agreed that can be a problem. However, I am not that familiar with the code so you'll have to guide me a bit.

I found this code snippet, and I assume what you mean is we need to a separate field for `TransactionType`:"payment" etc.

```
export interface TransactionMapping {
  Description?: TransactionDescriptionComponent
  Simple: TransactionSimpleComponent
  TableDetail?: TransactionTableDetailComponent
  parser: TransactionParser
  TransactionType: string          //  do this and set in each transaction type? 
}

```

mvadari replied:
Yes, something like that - though I'd recommend calling it something like `TransactionCategory`, since `TransactionType` is usually used for the type of transaction, e.g. Payment vs OfferCreate

TusharPardhe replied:
On it! Also, should I make these two fields optional or just create a component/function and return null in new transaction folders like OfferCancel, NFTokenMint etc.

```
Simple: TransactionSimpleComponent
parser: TransactionParser
```

mvadari replied:
The migration is still ongoing from the old transaction format to the new one. You can leave the original dictionary in for all the transactions that don't exist in this folder yet.

TusharPardhe replied:
So, what I understand is:

Make a folder in Transactions folder -> eg. OfferCreate
```
import { TransactionMapping } from '../types'
import { parser } from './parser'
import { Simple } from './Simple'

export const OfferCancelTransaction: TransactionMapping = {
  Simple,
  TransactionCategory: 'Dex',
  parser,
}

```
and two three files named index, simple and parser.

where simple is:
```
import React from 'react'
import { TransactionSimpleComponent, TransactionSimpleProps } from '../types'

const Simple: TransactionSimpleComponent = (props: TransactionSimpleProps) => {
  const { data } = props

  return <>{data}</>
}
export { Simple }

```

and parser is:

```
export function parser(tx: any, meta: any) {
  return tx
}

```

ckniffen replied:
Essentially that is the new setup.  Here is where they are defined and there is a README.

TusharPardhe replied:
Added all the tx types.

mvadari commented on src/containers/shared/components/Transaction/AccountDelete/Simple.tsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react'
```
Not all transactions have been migrated over to this format - they shouldn't be created like this, it won't work. You can use a dictionary/list for the ones that aren't migrated over yet.

TusharPardhe replied:
can you please elaborate

TusharPardhe replied:
@ckniffen commented last time that this would be fine. I created components for all the transactions, do I need to delete them and again create a dictionary for the ones that don't exist?

I think that'll create problems because then whenever someone will update a Transaction type he/she has to manually go and remove that key from the dictionary.

Either we should use dictionary for everything now and then remove it when all the transaction types have been migrated or create simple components for the ones that don't exist.

That's what I think will be better.

ckniffen replied:
Sorry.  I maybe worded it wrong.  There could be entry for the index.tsx of just the transaction grouping but not the components.  That way it will fall through to the old code until they are refactored.  The way you have it here will cause all of the not yet refactored transaction types to not render the right stuff.

TusharPardhe replied:
Ahh okay,
So now what I am doing is importing the transaction types from index.tsx (which are present) and for the rest (that are not yet migrated) defining them in the Ledger - index.ts file.


Like this:

    const TransactionCategories = {
      ...transactionTypes.TransactionCategory, // these are imported ones
      AccountDelete: 'Account', // these are which are yet to be modified
      AccountSet: 'Account',
      CheckCancel: 'Payment',
      CheckCash: 'Payment',
      CheckCreate: 'Payment',
      DepositPreauth: 'Account',
    }

Is this fine? @ckniffen @mvadari

mvadari replied:
An example of how we've done it previously is here: https://github.com/ripple/explorer/blob/master/src/containers/Transactions/Simple/index.js#L27-L34

TusharPardhe replied:
Thank you @mvadari that helped! I've changed the code.

mvadari commented on public/locales/en-US/translations.json:
```diff
@@ -160,6 +166,7 @@
   "transaction_EnableAmendment": "Enable Amendment",
   "transaction_SetFee": "Set Fee",
   "transaction_UNLModify": "UNL Modify",
+  "pseudo_transactions":"Pseudo Transactions",
```
nit: shouldn't this be with the categories above?

TusharPardhe replied:
fixed
